src/Microsoft.Graph.Core/Requests/RetryHandler.cs

```diff
@@ -0,0 +1,192 @@
+        /// </summary>
+        /// <param name="response">The <see cref="HttpResponseMessage"/>returned.</param>
+        /// <returns></returns>
+        public bool IsRetry(HttpResponseMessage response)
+        {
+            if ((response.StatusCode == HttpStatusCode.ServiceUnavailable ||
+                response.StatusCode == (HttpStatusCode)429))
+            {
+                return true;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Check the HTTP request's content to determine whether it can be retried or not.
+        /// </summary>
+        /// <param name="request">The <see cref="HttpRequestMessage"/>needs to be sent.</param>
+        /// <returns></returns>
+        private bool IsBuffered(HttpRequestMessage request)
+        {
+            HttpContent content = request.Content;
+
+            if ((request.Method == HttpMethod.Put || request.Method == HttpMethod.Post || request.Method.Method.Equals("PATCH")) 
+                && content != null && (content.Headers.ContentLength == null || (int)content.Headers.ContentLength == -1))
+            {
+                return false;
+            }
+            return true;
+           
+        }
+
+        /// <summary>
+        /// Update Retry-Attempt header in the HTTP request
+        /// </summary>
+        /// <param name="request">The <see cref="HttpRequestMessage"/>needs to be sent.</param>
+        /// <param name="retry_count">Retry times</param>
+        private void AddOrUpdateRetryAttempt(HttpRequestMessage request, int retry_count)
+        {
+            if (request.Headers.Contains(RETRY_ATTEMPT))
+            {
+                request.Headers.Remove(RETRY_ATTEMPT);
+            }
+            request.Headers.Add(RETRY_ATTEMPT, retry_count.ToString());
+        }
+
+        /// <summary>
+        /// Delay task operation based on Retry-After header in the response or exponential backoff
+        /// </summary>
+        /// <param name="response">The <see cref="HttpResponseMessage"/>returned.</param>
+        /// <param name="retry_count">The retry counts</param>
+        /// <param name="cancellationToken">The cancellationToken for the Http request</param>
+        /// <returns>The <see cref="Task"/> for delay operation.</returns>
+        public Task Delay(HttpResponseMessage response, int retry_count, CancellationToken cancellationToken)
+        {
+            
+            TimeSpan delay = TimeSpan.FromMilliseconds(0);
+            HttpHeaders headers = response.Headers;
+            if (headers.TryGetValues(RETRY_AFTER, out IEnumerable<string> values))
+            {
+                string retry_after = values.First();    
+                if (Int32.TryParse(retry_after, out int delay_seconds))
+                {
+                    delay = TimeSpan.FromSeconds(delay_seconds);
+                }
+            }
+            else
+            {
+
```

Comment:
At retry_count == 10, m_pow is 1024, delay_time is 10240000 ms, which is a 170 minute final backoff, correct? If so, do we want a final backoff of 170 minutes? I don't know the answer here so I wonder if there was discussion about what max backoff should be. We did remove the max backoff const.

Reply:
Yes. It is how this calculation. Darrel may want to give enough time for services to recover and if time becomes too larger and the services cannot return successful responses, usually it will be returned with timeout errors. Retry may need to catch timeout in future, this still needs to be discussed.